exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/BaseParquetMetadataProvider.java

```diff
@@ -103,34 +110,39 @@
   // whether metadata for row groups should be collected to create files, partitions and table metadata
   private final boolean collectMetadata = false;
 
-  public BaseParquetMetadataProvider(List<ReadEntryWithPath> entries,
-                                     ParquetReaderConfig readerConfig,
-                                     String tableName,
-                                     Path tableLocation,
-                                     TupleMetadata schema,
-                                     DrillStatsTable statsTable) {
-    this(readerConfig, entries, tableName, tableLocation, schema, statsTable);
-  }
+  protected BaseParquetMetadataProvider(Builder<?> builder) {
+    if (builder.entries != null) {
+      // reuse previously stored metadata
+      this.entries = builder.entries;
+      this.tableName = builder.selectionRoot != null ? builder.selectionRoot.toUri().getPath() : "";
+      this.tableLocation = builder.selectionRoot;
+    } else if (builder.selection != null) {
+      this.entries = new ArrayList<>();
+      this.tableName = builder.selection.getSelectionRoot() != null ? builder.selection.getSelectionRoot().toUri().getPath() : "";
+      this.tableLocation = builder.selection.getSelectionRoot();
+    } else {
+      // case of hive parquet table
+      this.entries = new ArrayList<>();
+      this.tableName = null;
+      this.tableLocation = null;
+    }
 
-  public BaseParquetMetadataProvider(ParquetReaderConfig readerConfig,
-                                     List<ReadEntryWithPath> entries,
-                                     String tableName,
-                                     Path tableLocation,
-                                     TupleMetadata schema,
-                                     DrillStatsTable statsTable) {
-    this.entries = entries == null ? new ArrayList<>() : entries;
-    this.readerConfig = readerConfig == null ? ParquetReaderConfig.getDefaultInstance() : readerConfig;
-    this.tableName = tableName;
-    this.tableLocation = tableLocation;
-    this.schema = schema;
-    this.statsTable = statsTable;
-  }
+    SchemaProvider schemaProvider = builder.metadataProviderManager.getSchemaProvider();
+    TupleMetadata schema = builder.schema;
+    // schema passed into the builder has greater priority
+    if (schema == null && schemaProvider != null) {
+      try {
+        schema = schemaProvider.read().getSchema();
```

Comment:
What is the intent of the schema here? We want to plan a query. Historically, the schema is entirely dynamic: provided at run time. With the provided schema, the user can constrain that dynamic schema to use the provided schema definition (which can be strict or lenient.)

With metadata, we are feeding the scan's "exhaust" (dynamically-discovered schema) back to itself as part of the plan.

What do we expect the can do do with this schema? Ignore and continue to produce a dynamic schema (that is, discover any new columns, omit any old columns which have disappeared as the last of old files were deleted, etc.)

Or, do we expect the scan to adjust its new scans to use the schema from prior scans? There is nothing in this PR to enforce such as schema. (As it turns out, that is what I'm currently doing.)

So, if this PR has no way to use the schema, do we even care about it? Do we need to even fetch it? Or, is the schema here needed to understand the available metadata for partition pruning?

Reply:
Please explain a bit more. At present, the `EasySubScan` has a `schema` attribute that contains a provided schema. Does the code here choose which schema appears in that field? Or, does the code here change that value?

What I'm trying to understand is this: we now have three ways to specify schema: 1) Metastore, 2) Provided schema file, 3) Table function. (Just reviewed the documentation PR that describe this - very helpful.) What is the precedence? I'd think table functions are first. But, does that schema have to agree with the other schemas? Does the planner check?

Can I have both a provided and metadata schema? If I have a provided schema, will the metadata system (`ANALYZE TABLE`) use it?

How do those precedence rules affect the builder precedence here?

I'm working on cleaning up scan-time schema usage, so I need to understand how you want all this to work.

Reply:
The highest priority has the inline schema. Also, the schema file may be provided in a table function, but it cannot be used with the inline schema for the same table within the same query. If the table function wasn't specified, but the schema file is present and `store.table.use_schema_file` was enabled, the schema file will be used.
In the opposite case, metastore schema will be used. Only a single schema source may be used.

Yes, `ANALYZE TABLE` will use the provided schema.

In this builder, the `schema` field - is a target resolved schema after group scan deserialization. For the case when the schema is null - will be used `SchemaProvider` which provides schema either from table function or from schema file. If the schema is still null, metastore schema wouldn't be overridden and will be used.